File: update_security_groups_lambda/update_security_groups.py
'''
Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.


Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at


    http://aws.amazon.com/apache2.0/


or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
'''

# Ports your application uses that need inbound permissions from the service for
INGRESS_PORTS = { 'http' : 80, 'https': 443, 'example': 8080 }
# Tags which identify the security groups you want to update
# For a group to be updated it will need to have 3 properties that are true:
# 1. It has to be tagged 'Protocol: X' (Where 'X' is one of your INGRESS_PORTS above)
# 2. It has to be tagged 'Name: cloudfront_g' or 'Name: cloudfront_r'
# 3. It has to be tagged 'AutoUpdate: true'
# If any of these 3 are not true, the security group will be unmodified.
GLOBAL_SG_TAGS = { 'Name': 'cloudfront_g', 'AutoUpdate': 'true' }
REGION_SG_TAGS = { 'Name': 'cloudfront_r', 'AutoUpdate': 'true' }

import boto3
import hashlib
import json
import logging
import urllib.request, urllib.error, urllib.parse
import os

def lambda_handler(event, context):
    global logger
    logger.setLevel(logging.CRITICAL)

    # Set the environment variable DEBUG to 1 if you want verbose debug details in CloudWatch Logs.
    try:
        if bool(strtobool(os.environ.get('DEBUG', ''))):
            logger.setLevel(logging.DEBUG)
    except ValueError:
        pass
    

    # If you want a different service, set the SERVICE environment variable.
    # It defaults to CLOUDFRONT. Using 'jq' and 'curl' get the list of possible
    # services like this:
    # curl -s 'https://ip-ranges.amazonaws.com/ip-ranges.json' | jq -r '.prefixes[] | .service' ip-ranges.json | sort -u 
    SERVICE = os.getenv( 'SERVICE', "CLOUDFRONT")
    
    logger.debug(("Received event: " + json.dumps(event, indent=2)))
    message = json.loads(event['Records'][0]['Sns']['Message'])

    # Load the ip ranges from the url
    ip_ranges = json.loads(get_ip_groups_json(message['url'], message['md5']))

    # Extract the service ranges
    global_cf_ranges = get_ranges_for_service(ip_ranges, SERVICE, "GLOBAL")
    region_cf_ranges = get_ranges_for_service(ip_ranges, SERVICE, "REGION")

    # Update the security groups
    result = update_security_groups(global_cf_ranges, "GLOBAL")
    result = result + update_security_groups(global_cf_ranges, "REGION")
    
    return result


def get_ip_groups_json(url, expected_hash):
    global logger

    logger.debug("Updating from " + url)

    response = urllib.request.urlopen(url)
    ip_json = response.read()

    m = hashlib.md5()
    m.update(ip_json)
    hash = m.hexdigest()

    if hash != expected_hash:
        raise Exception('MD5 Mismatch: got ' + hash + ' expected ' + expected_hash)

    return ip_json

def get_ranges_for_service(ranges, service, subset):
    global logger

    service_ranges = list()
    for prefix in ranges['prefixes']:
        if prefix['service'] == service and ((subset == prefix['region'] and subset == "GLOBAL") or (subset != 'GLOBAL' and prefix['region'] != 'GLOBAL')):
            logger.debug(('Found ' + service + ' region: ' + prefix['region'] + ' range: ' + prefix['ip_prefix']))
            service_ranges.append(prefix['ip_prefix'])

    return service_ranges

def update_security_groups(new_ranges, rangeType):
    global logger

    client = boto3.client('ec2')
    
    # All the security groups we will need to find.
    allSGs = INGRESS_PORTS.keys()
    # Iterate over every group, doing its global and regional versions
    for curGroup in allSGs:
        tagToFind = {}
        if rangeType == "GLOBAL":
            tagToFind = GLOBAL_SG_TAGS
        else:
            tagToFind = REGION_SG_TAGS    
        tagToFind['Protocol'] = curGroup
        rangeToUpdate = get_security_groups_for_update(client, tagToFind)        
        logger.debug('Found {} groups tagged {}, proto {} to update'.format(
                str(len(rangeToUpdate)),
                tagToFind["Name"],
                curGroup))

        result = list()
        
        if update_security_group(client, rangeToUpdate[0], new_ranges, INGRESS_PORTS[curGroup] ):
            result.append('Updated ' + rangeToUpdate[0]['GroupId'])

    return result


def update_security_group(client, group, new_ranges, port):
    added = 0
    removed = 0
    global logger

    if len(group['IpPermissions']) > 0:
        for permission in group['IpPermissions']:
            if permission['FromPort'] <= port and permission['ToPort'] >= port:
                old_prefixes = list()
                to_revoke = list()
                to_add = list()
                for range in permission['IpRanges']:
                    cidr = range['CidrIp']
                    old_prefixes.append(cidr)
                    if new_ranges.count(cidr) == 0:
                        to_revoke.append(range)
                        logger.debug((group['GroupId'] + ": Revoking " + cidr + ":" + str(permission['ToPort'])))

                for range in new_ranges:
                    if old_prefixes.count(range) == 0:
                        to_add.append({ 'CidrIp': range })
                        logger.debug((group['GroupId'] + ": Adding " + range + ":" + str(permission['ToPort'])))

                removed += revoke_permissions(client, group, permission, to_revoke)
                added += add_permissions(client, group, permission, to_add)
    else:
        to_add = list()
        for range in new_ranges:
            to_add.append({ 'CidrIp': range })
            logger.debug((group['GroupId'] + ": Adding " + range + ":" + str(port)))
        permission = { 'ToPort': port, 'FromPort': port, 'IpProtocol': 'tcp'}
        added += add_permissions(client, group, permission, to_add)

    logger.debug((group['GroupId'] + ": Added " + str(added) + ", Revoked " + str(removed)))
    return (added > 0 or removed > 0)


def revoke_permissions(client, group, permission, to_revoke):
    if len(to_revoke) > 0:
        revoke_params = {
            'ToPort': permission['ToPort'],
            'FromPort': permission['FromPort'],
            'IpRanges': to_revoke,
            'IpProtocol': permission['IpProtocol']
        }

        client.revoke_security_group_ingress(GroupId=group['GroupId'], IpPermissions=[revoke_params])

    return len(to_revoke)


def add_permissions(client, group, permission, to_add):
    if len(to_add) > 0:
        add_params = {
            'ToPort': permission['ToPort'],
            'FromPort': permission['FromPort'],
            'IpRanges': to_add,
            'IpProtocol': permission['IpProtocol']
        }

        client.authorize_security_group_ingress(GroupId=group['GroupId'], IpPermissions=[add_params])

    return len(to_add)


def get_security_groups_for_update(client, security_group_tag):
    filters = list()
    for key, value in security_group_tag.items():
        filters.extend(
            [
                { 'Name': "tag-key", 'Values': [ key ] },
                { 'Name': "tag-value", 'Values': [ value ] }
            ]
        )

    response = client.describe_security_groups(Filters=filters)

    return response['SecurityGroups']

# This is a handy test event you can use when testing your lambda function.
'''
Sample Event From SNS:

{
  "Records": [
    {
      "EventVersion": "1.0",
      "EventSubscriptionArn": "arn:aws:sns:EXAMPLE",
      "EventSource": "aws:sns",
      "Sns": {
        "SignatureVersion": "1",
        "Timestamp": "1970-01-01T00:00:00.000Z",
        "Signature": "EXAMPLE",
        "SigningCertUrl": "EXAMPLE",
        "MessageId": "95df01b4-ee98-5cb9-9903-4c221d41eb5e",
        "Message": "{\"create-time\": \"yyyy-mm-ddThh:mm:ss+00:00\", \"synctoken\": \"0123456789\", \"md5\": \"45be1ba64fe83acb7ef247bccbc45704\", \"url\": \"https://ip-ranges.amazonaws.com/ip-ranges.json\"}",
        "Type": "Notification",
        "UnsubscribeUrl": "EXAMPLE",
        "TopicArn": "arn:aws:sns:EXAMPLE",
        "Subject": "TestInvoke"
      }
    }
  ]
}

'''

File: update_security_groups_lambda/update_security_groups.js
// Lambda function for updating the cloudfront EC2 security group ingress rules with the CloudFront IP range changes using the Node.js 8.10 runtime.
//
// Name of the service, as seen in the ip-groups.json file, to extract information for
const SERVICE = 'CLOUDFRONT';
// Ports your application uses that need inbound permissions from the service for
const INGRESS_PORTS = { Http: 80, Https: 443 };
// Tags which identify the security groups you want to update
const SECURITY_GROUP_TAG_FOR_GLOBAL_HTTP = { Name: 'cloudfront_g', AutoUpdate: 'true', Protocol: 'http' };
const SECURITY_GROUP_TAG_FOR_GLOBAL_HTTPS = { Name: 'cloudfront_g', AutoUpdate: 'true', Protocol: 'https' };
const SECURITY_GROUP_TAG_FOR_REGION_HTTP = { Name: 'cloudfront_r', AutoUpdate: 'true', Protocol: 'http' };
const SECURITY_GROUP_TAG_FOR_REGION_HTTPS = { Name: 'cloudfront_r', AutoUpdate: 'true', Protocol: 'https' };
const REGION = process.env.REGION; // requires REGION environment variable to be set
const request = require('request');
const crypto = require('crypto');
const AWS = require('aws-sdk');

module.exports = async function(event, context, callback) {
  console.log('Received event: ' + JSON.stringify(event, null, 2));
  let message = JSON.parse(event.Records[0].Sns.Message);
  //  Load the ip ranges from the url
  let ipRanges = await getIpGroupsJSON(message.url, message.md5).catch((err) => { console.log(err); return callback(err); });

  // extract the service ranges
  let globalCFRanges = getRangesForService(ipRanges, SERVICE, 'GLOBAL');
  let regionCFRanges = getRangesForService(ipRanges, SERVICE, 'REGION');
  ipRanges = { GLOBAL: globalCFRanges, REGION: regionCFRanges };

  // update the security groups
  let result = await updateSecurityGroups(ipRanges).catch((err) => { console.log(err); return callback(err); });

  console.log('Update results:', result);
  return callback(null, result);
};

function getIpGroupsJSON(url, expectedHash) {
  return new Promise((resolve, reject) => {
    console.log('Updating from ' + url);
    request.get(url, function(err, resp, body) {
      if (err) {
        console.log(`Encountered error when getting url: ${url} Err: ${err}`);
        return reject(err);
      }
      if (resp.statusCode !== 200) {
        console.log(`Non 200 status code when getting url: ${url} statusCode: ${resp.statusCode}, body: ${JSON.stringify(body)}`);
        return reject(new Error(`Non-200 status code when getting url: ${url} statusCode: ${resp.statusCode}, body: ${JSON.stringify(body)}`));
      }
      let ipJSON = JSON.parse(body);
      let hash = crypto.createHash('md5').update(JSON.stringify(ipJSON)).digest('hex');
      if (hash !== expectedHash) {
        return reject(new Error('MD5 Mismatch: got ' + hash + ' expected ' + expectedHash));
      }
      return resolve(ipJSON);
    });
  });
}

function getRangesForService(ranges, service, subset) {
  let serviceRanges = [];
  for (let prefix of ranges.prefixes) {
    if (prefix.service === service &&
      ((subset === prefix.region && subset === 'GLOBAL') || (subset !== 'GLOBAL' && prefix.region !== 'GLOBAL'))
    ) {
      console.log('Found ' + service + ' region: ' + prefix['region'] + ' range: ' + prefix['ip_prefix']);
      serviceRanges.push(prefix['ip_prefix']);
    }
  }
  return serviceRanges;
}

async function updateSecurityGroups(newRanges) {
  let ec2 = new AWS.EC2({ region: REGION });
  let globalHttpGroup = await getSecurityGroupsForUpdate(ec2, SECURITY_GROUP_TAG_FOR_GLOBAL_HTTP).catch((err) => { throw new Error(err); });
  let globalHttpsGroup = await getSecurityGroupsForUpdate(ec2, SECURITY_GROUP_TAG_FOR_GLOBAL_HTTPS).catch((err) => { throw new Error(err); });
  let regionHttpGroup = await getSecurityGroupsForUpdate(ec2, SECURITY_GROUP_TAG_FOR_REGION_HTTP).catch((err) => { throw new Error(err); });
  let regionHttpsGroup = await getSecurityGroupsForUpdate(ec2, SECURITY_GROUP_TAG_FOR_REGION_HTTPS).catch((err) => { throw new Error(err); });


  console.log(`Found ${JSON.stringify(globalHttpGroup)} Cloudfront_g HttpSecurityGroups to update`);
  console.log(`Found ${JSON.stringify(globalHttpsGroup)} Cloudfront_g HttpsSecurityGroups to update`);
  console.log(`Found ${JSON.stringify(globalHttpGroup)} Cloudfront_r HttpSecurityGroups to update`);
  console.log(`Found ${JSON.stringify(globalHttpsGroup)} Cloudfront_r HttpsSecurityGroups to update`);
  let result = [];

  let globalHttpUpdated = 0;
  let globalHttpsUpdated = 0;
  let regionHttpUpdated = 0;
  let regionHttpsUpdated = 0;

  if (globalHttpGroup && globalHttpGroup.length) {
    for (let group in globalHttpGroup) {
      let updatedThisGroup = await updateSecurityGroup(ec2, globalHttpGroup[group], newRanges['GLOBAL'], INGRESS_PORTS['Http']).catch((err) => { throw new Error(err); });
      if (updatedThisGroup) {
        globalHttpUpdated += 1;
        result.push(`Updated ${globalHttpGroup[group].GroupId}`);
      }
    }
  }
  if (globalHttpsGroup && globalHttpsGroup.length) {
    for (let group in globalHttpsGroup) {
      let updatedThisGroup = await updateSecurityGroup(ec2, globalHttpsGroup[group], newRanges['GLOBAL'], INGRESS_PORTS['Https']).catch((err) => { throw new Error(err); });
      if (updatedThisGroup) {
        globalHttpsUpdated += 1;
        result.push(`Updated ${globalHttpsGroup[group].GroupId}`);
      }
    }
  }
  if (regionHttpGroup && regionHttpGroup.length) {
    for (let group in regionHttpGroup) {
      let updatedThisGroup = await updateSecurityGroup(ec2, regionHttpGroup[group], newRanges['REGION'], INGRESS_PORTS['Http']).catch((err) => { throw new Error(err); });
      if (updatedThisGroup) {
        regionHttpUpdated += 1;
        result.push(`Updated ${regionHttpGroup[group].GroupId}`);
      }
    }
  }
  if (regionHttpsGroup && regionHttpsGroup.length) {
    for (let group in regionHttpsGroup) {
      let updatedThisGroup = await updateSecurityGroup(ec2, regionHttpsGroup[group], newRanges['REGION'], INGRESS_PORTS['Https']).catch((err) => { throw new Error(err); });
      if (updatedThisGroup) {
        regionHttpsUpdated += 1;
        result.push(`Updated ${regionHttpsGroup[group].GroupId}`);
      }
    }
  }

  result.push(`Updated ${globalHttpUpdated} of ${JSON.stringify(globalHttpGroup)} CloudFront_g HttpSecurityGroups`);
  result.push(`Updated ${globalHttpsUpdated} of ${JSON.stringify(globalHttpsGroup)} CloudFront_g HttpsSecurityGroups`);
  result.push(`Updated ${regionHttpUpdated} of ${JSON.stringify(regionHttpGroup)} CloudFront_r HttpSecurityGroups`);
  result.push(`Updated ${regionHttpsUpdated} of ${JSON.stringify(regionHttpsGroup)} CloudFront_r HttpsSecurityGroups`);

  return result;
}

function countOccurence(array, item) {
  let count = 0;
  for (let i = 0; i < array.length; i++) {
    if (array[i] === item) {
      count += 1;
    }
  }
  return count;
}

async function updateSecurityGroup(client, group, newRanges, port) {
  let added = 0;
  let removed = 0;
  if (group.IpPermissions && group.IpPermissions.length) {
    for (let p in group.IpPermissions) {
      let permission = group.IpPermissions[p];
      if (permission.FromPort <= port && permission.ToPort >= port) {
        let oldPrefixes = [];
        let toRevoke = [];
        let toAdd = [];
        for (let r in permission.IpRanges) {
          let range = permission.IpRanges[r];
          let cidr = range.CidrIp;
          oldPrefixes.push(cidr);
          if (countOccurence(newRanges, cidr) === 0) {
            toRevoke.push(range);
            console.log(group.GroupId + ': Revoking ' + cidr + ':' + permission['ToPort'].toString());
          }
        }

        for (let r in newRanges) {
          let range = newRanges[r];
          if (countOccurence(oldPrefixes, range) === 0) {
            toAdd.push({ CidrIp: range });
            console.log(group['GroupId'] + ': Adding ' + range + ':' + permission['ToPort'].toString());
          }
        }

        removed += await revokePermissions(client, group, permission, toRevoke).catch((err) => { throw new Error(err); });

        added += await addPermissions(client, group, permission, toAdd).catch((err) => { throw new Error(err); });

      }
    }
  } else {
    let toAdd = [];
    for (let r in newRanges) {
      let range = newRanges[r];
      toAdd.push({ CidrIp: range });
      console.log(group['GroupId'] + ': Adding ' + range + ':' + port);
    }
    let permission = { ToPort: port, FromPort: port, IpProtocol: 'tcp' };
    added += await addPermissions(client, group, permission, toAdd).catch((err) => { throw new Error(err); });
  }

  console.log(group['GroupId'] + ': Added ' + added.toString() + ', Revoked ' + removed.toString());
  return (added > 0 || removed > 0);
}

function revokePermissions(client, group, permission, toRevoke) {
  return new Promise((resolve, reject) => {
    if (!toRevoke.length) {
      return resolve(toRevoke.length);
    }
    let revokeParams = {
      ToPort: permission['ToPort'],
      FromPort: permission['FromPort'],
      IpRanges: toRevoke,
      IpProtocol: permission['IpProtocol']
    };
    client.revokeSecurityGroupIngress({
      GroupId: group.GroupId,
      IpPermissions: [revokeParams]
    }, function(err) {
      if (err) {
        return reject(err);
      }
      return resolve(toRevoke.length);
    });
  });
}

function addPermissions(client, group, permission, toAdd) {
  return new Promise((resolve, reject) => {
    if (!toAdd.length) {
      return resolve(toAdd.length);
    }
    let addParams = {
      ToPort: permission['ToPort'],
      FromPort: permission['FromPort'],
      IpRanges: toAdd,
      IpProtocol: permission['IpProtocol']
    };
    client.authorizeSecurityGroupIngress({
      GroupId: group.GroupId,
      IpPermissions: [addParams]
    }, function(err) {
      if (err) {
        return reject(err);
      }
      return resolve(toAdd.length);
    });
  });
}

function getSecurityGroupsForUpdate(client, securityGroupTag) {
  let filters = [];
  for (let i in securityGroupTag) {
    filters.push({
      Name: 'tag-key',
      Values: [i]
    });
    filters.push({
      Name: 'tag-value',
      Values: [securityGroupTag[i]]
    });
  }
  return new Promise((resolve, reject) => {
    client.describeSecurityGroups({
      Filters: filters
    }, function(err, response) {
      if (err) {
        return reject(err);
      }
      return resolve(response.SecurityGroups);
    });
  });
}
